### PR TITLE
gputest.py: Add dawn end2end tests expectations

### PR DESCRIPTION
**Pull request by haoxli**
- Rename expectationhelper.py to testhelp.py
- Refactor Util.get_test_result() and move to testhelp.py
- Add function to move known failures from pass_fail num to fail_fail num for dawn end2end tests.

### CODE REVIEW

**Comment by haoxli** [comment]
@gyagp  PTAL, thanks!

File: misc/testhelper.py [hunk 1]
```diff
@@ -123,3 +126,101 @@ def update(target, root_dir):
...
+            for new_key, new_val in val.items():
+                self._parse_result(new_key, new_val, '%s/%s' % (path, new_key))
+
+    def count_pass_fail_to_fail_fail(self):
+        self.fail_fail.extend(self.pass_fail)
+        self.pass_fail.clear()
+
+    def update_dawn_expectations(self, target):
+        expectations = TestExpectation.LOCAL_EXPECTATIONS.get(target)
+        if expectations is None or len(expectations) == 0:
```
Comment (gyagp) [inline misc/testhelper.py:201]: if not expectations?

File: misc/testhelper.py [hunk 2]
```diff
@@ -123,3 +126,101 @@ def update(target, root_dir):
...
+                self.fail_pass.append(path)
+            elif expected_pass and not actual_pass:
+                self.pass_fail.append(path)
+            elif expected_pass and actual_pass:
+                self.pass_pass.append(path)
+        else:
+            for new_key, new_val in val.items():
+                self._parse_result(new_key, new_val, '%s/%s' % (path, new_key))
+
+    def count_pass_fail_to_fail_fail(self):
```
Comment (gyagp) [inline misc/testhelper.py:195]: This is not a general usage. Please remove this function, and unfold the logic to dawn_end2end_tests_runsuppressed.

File: misc/testhelper.py [hunk 3]
```diff
@@ -123,3 +126,101 @@ def update(target, root_dir):
...
+        for test in self.pass_fail:
+            expectation = f'[ {Util.HOST_OS} ] {test}'
+            if expectation in expectations:
+                expected_fail.append(test)
+        for test in expected_fail:
+            self.fail_fail.append(test)
+            self.pass_fail.remove(test)
+
+    @property
+    def pass_fail_num(self):
```
Comment (gyagp) [inline misc/testhelper.py:213]: These functions don't save a lot. I suggest to remove them and use len() instead.

File: misc/testhelper.py [hunk 4]
```diff
@@ -123,3 +126,101 @@ def update(target, root_dir):
...
+
+    def count_pass_fail_to_fail_fail(self):
+        self.fail_fail.extend(self.pass_fail)
+        self.pass_fail.clear()
+
+    def update_dawn_expectations(self, target):
+        expectations = TestExpectation.LOCAL_EXPECTATIONS.get(target)
+        if expectations is None or len(expectations) == 0:
+            return
+        expected_fail = []
```
Comment (gyagp) [inline misc/testhelper.py:203]: Remove expected_fail and simplify the two loops into one.

File: misc/testhelper.py [hunk 5]
```diff
@@ -123,3 +126,101 @@ def update(target, root_dir):
...
+                self.pass_pass.append(path)
+        else:
+            for new_key, new_val in val.items():
+                self._parse_result(new_key, new_val, '%s/%s' % (path, new_key))
+
+    def count_pass_fail_to_fail_fail(self):
+        self.fail_fail.extend(self.pass_fail)
+        self.pass_fail.clear()
+
+    def update_dawn_expectations(self, target):
```
Comment (gyagp) [inline misc/testhelper.py:199]: Remove this function and fold the logic into dawn_e2e_tests.

**Review by gyagp — APPROVED** [review]
LGTM with one nit

File: misc/gputest.py [hunk 6]
```diff
@@ -508,25 +508,36 @@ def _report(self):
...
+                elif re.search('dawn_end2end_tests', op):
+                    expectations = TestExpectation.LOCAL_EXPECTATIONS.get(op.split('-')[1])
+                    if expectations:
+                        pass_fail = result.pass_fail.copy()
+                        for test in pass_fail:
+                            expectation = f'[ {Util.HOST_OS} ] {test}'
+                            if expectation in expectations:
+                                result.fail_fail.append(test)
+                                result.pass_fail.remove(test)
+
```
[the remark this answers is not in the document]
Comment (haoxli) [reply misc/gputest.py:527]: remove the blank line and use reverse order loop to save one copy.

**Review by gyagp — APPROVED** [review]

**Pull request merged**